icu4c/source/common/rbbirb.cpp

```diff
@@ -148,13 +149,19 @@ RBBIDataHeader *RBBIRuleBuilder::flattenData() {
     //   Sizes here are padded up to a multiple of 8 for better memory alignment.
     //   Sections sizes actually stored in the header are for the actual data
     //     without the padding.
-    //
+
     int32_t headerSize        = align8(sizeof(RBBIDataHeader));
     int32_t forwardTableSize  = align8(fForwardTable->getTableSize());
     int32_t reverseTableSize  = align8(fForwardTable->getSafeTableSize());
     int32_t trieSize          = align8(fSetBuilder->getTrieSize());
     int32_t statusTableSize   = align8(fRuleStatusVals->size() * sizeof(int32_t));
-    int32_t rulesSize         = align8((fStrippedRules.length()+1) * sizeof(UChar));
+
+    int32_t rulesLengthInUTF8 = 0;
+    u_strToUTF8WithSub(0, 0, &rulesLengthInUTF8,
+                       fStrippedRules.getTerminatedBuffer(), -1, 0xfffd, nullptr, fStatus);
```

Comment:
`fStrippedRules.getTerminatedBuffer(), -1` --> `fStrippedRules.getBuffer(), fStrippedRules.length()`

And since you rely on NUL-termination, I would check somewhere that `fStrippedRules.indexOf(u'\0') < 0` and set an illegal-argument-error code.
Alternatively, allow NUL, store the length (bytes, not counting the NUL) in the data file, and use it when loading the data. (That would also save you a strlen() at load time.)